File: ch01/examples/example-1.4.1.6/Main.hs
module Main where

main :: IO ()
main = do
  print "Result from formula: "
  print $ sineProduct 100_000 0.1
  print "Result from sine function: "
  print sineResult

sineProduct n x = product $ map (\k -> 1.0 - x * x / k / k) [1 .. n]

sineResult = sin (pi * 0.1) / (pi * 0.1)

sineProductResult = sineProduct 100_000 0.1

File: ch01/examples/example-1.4.1.6/main.ml
let sine_product (n : int) (x : float) : float =
  let open Float in
  List.init n (fun k -> k + 1)
  |> List.map (fun k -> 1.0 -. (x *. x /. (of_int k *. of_int k)))
  |> List.fold_left ( *. ) 1.0

let () =
  let open Printf in
  let sine_result = sin (Float.pi *. 0.1) /. (Float.pi *. 0.1) in

  printf "Result from formula: %f\n" (sine_product 100_000 0.1);
  printf "Result from sine function: %f\n" sine_result
